build_modules/lib/src/meta_module_clean_builder.dart

```diff
@@ -0,0 +1,125 @@
+// Copyright (c) 2018, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:async';
+import 'dart:collection';
+import 'dart:convert';
+
+import 'package:graphs/graphs.dart';
+
+import 'package:build/build.dart';
+import 'meta_module.dart';
+import 'meta_module_builder.dart';
+import 'modules.dart';
+
+/// The extension for serialized clean meta module assets.
+const metaModuleCleanExtension = '.meta_module_clean';
+
+/// Map of [AssetId] to corresponding primary [AssetId] within the same
+/// [Module].
+final _assetToModule =
+    new Resource<Map<AssetId, Module>>(() => {}, dispose: (map) => map.clear());
+
+/// Returns a set of all modules transitvely reachable from the provided meta
+/// module asset.
+Future<Set<Module>> _transitiveModules(BuildStep buildStep, AssetId metaAsset,
```

Comment:
I wonder how well this fits with `crawlAsync`?

https://pub.dartlang.org/documentation/graphs/0.1.1/graphs/crawlAsync.html

The benefit would be that we'd separate out the business logic of how we resolve modules and dependencies from the graph algorithm

Reply:
You could do a `crawlAsync` for `MetaModule`s (children would be all the merged MetaModule dependencies of all the contained modules), and then you could expand all the modules afterwords?

Reply:
I can't use the crawlAsync method since it doesn't propagate errors as expected.

Reply:
https://github.com/dart-lang/graphs/pull/18